Fixes #1630: Tools to explore the issues of operation response timing

KSchopmeyer opened the pull request:
This modifies a test and develops a second test to allow us to
explore the issues of response time documented in issue #1630

1. It adds more detailed tests to the manual test  (tests/manual/run_enum_performance.py erformance and a better output display. It also adds command line argument processing so test variations can be defined.  This tool still works only with OpenPegasus because it uses a special OpenPegasus provider to generate the responses.

2. Adds a new test script response_performance.py (right now located in tests/performance) that
creates an xml enumerateinstances response of varying size instance and varying number of nstances returned and then executes the TupleParser.

This is a command line script with input parameters to control  the test.  It allows the user to select a variety of:

1. number of instances in the response

2. Size of the response in bytes.

3. Choice of running a test with no profiler running or with either the pyinstrument (call stack profiler) or the cProfile.

A table showing the response time for each combination of number of instances and size of the instances is displayed and if one of the profilers is defined by the input parameter, the
results of that profiler is displayed.

In addition, if the cProfiler is called it writes a file of the profile statistics to a file cprofiler.profile that can be explored with the stats module by executing:

   python -m python -m pstats <Name of file with profile data>

DISCUSSION:   

1. Is this useful  in finding the issues in response time?

2. Is there more tooling that would help?

andy-maier commented on tests/performance/response_performance.py:
```diff
@@ -0,0 +1,1304 @@
+<VALUE.NAMEDINSTANCE><INSTANCENAME CLASSNAME="TST_ResponseStressTestCxx"><KEYBINDING NAME="Id"><KEYVALUE VALUETYPE="string">83</KEYVALUE></KEYBINDING></INSTANCENAME><INSTANCE CLASSNAME="TST_ResponseStressTestCxx" ><PROPERTY NAME="Id"  TYPE="string"><VALUE>83</VALUE></PROPERTY><PROPERTY NAME="SequenceNumber"  TYPE="uint64"><VALUE>83</VALUE></PROPERTY><PROPERTY NAME="Pattern"  TYPE="string"><VALUE> ...
+<VALUE.NAMEDINSTANCE><INSTANCENAME CLASSNAME="TST_ResponseStressTestCxx"><KEYBINDING NAME="Id"><KEYVALUE VALUETYPE="string">84</KEYVALUE></KEYBINDING></INSTANCENAME><INSTANCE CLASSNAME="TST_ResponseStressTestCxx" ><PROPERTY NAME="Id"  TYPE="string"><VALUE>84</VALUE></PROPERTY><PROPERTY NAME="SequenceNumber"  TYPE="uint64"><VALUE>84</VALUE></PROPERTY><PROPERTY NAME="Pattern"  TYPE="string"><VALUE> ...
+<VALUE.NAMEDINSTANCE><INSTANCENAME CLASSNAME="TST_ResponseStressTestCxx"><KEYBINDING NAME="Id"><KEYVALUE VALUETYPE="string">85</KEYVALUE></KEYBINDING></INSTANCENAME><INSTANCE CLASSNAME="TST_ResponseStressTestCxx" ><PROPERTY NAME="Id"  TYPE="string"><VALUE>85</VALUE></PROPERTY><PROPERTY NAME="SequenceNumber"  TYPE="uint64"><VALUE>85</VALUE></PROPERTY><PROPERTY NAME="Pattern"  TYPE="string"><VALUE> ...
+<VALUE.NAMEDINSTANCE><INSTANCENAME CLASSNAME="TST_ResponseStressTestCxx"><KEYBINDING NAME="Id"><KEYVALUE VALUETYPE="string">86</KEYVALUE></KEYBINDING></INSTANCENAME><INSTANCE CLASSNAME="TST_ResponseStressTestCxx" ><PROPERTY NAME="Id"  TYPE="string"><VALUE>86</VALUE></PROPERTY><PROPERTY NAME="SequenceNumber"  TYPE="uint64"><VALUE>86</VALUE></PROPERTY><PROPERTY NAME="Pattern"  TYPE="string"><VALUE> ...
```
Seems the flake8 in py3.7 with minimum package levels does not recognize the underscore and its special meaning. Either remove the assignment, or find out which flake8 version has the support so we can upgrade the minimum version.

KSchopmeyer replied:
Removed the  _

andy-maier commented:
One more comment: The new packages pull in the following dependent packages. The new packages and their dependent packages should be added to the `minimum-constraints.txt` file, as follows:

```
# Performance profiling tools
pyinstrument==3.0.1
pyinstrument-cext==0.2.0  # from pyinstrument
tabulate==0.8.3
```

andy-maier commented:
Some more comments:
* usage line says "server"
* some formatting issues in --help
* remove the hard coded CIM-XML

andy-maier commented:
WHen redirecting the output of the stack profile, it fails as follows:

andy-maier commented:
One more issue:

When redirecting the output of a stack profile run, it fails as follows:

```
python tests/performance/response_performance.py -p stack -c 1000 -s 500 >perf_stack_1000_500_20190227094113.log
Traceback (most recent call last):
  File "tests/performance/response_performance.py", line 1304, in <module>
    main()
  File "tests/performance/response_performance.py", line 1291, in main
    table_rows = execute_tests(args)
  File "tests/performance/response_performance.py", line 1191, in execute_tests
    return execute_pyinstrument_tests(args)
  File "tests/performance/response_performance.py", line 1182, in execute_pyinstrument_tests
    print(profiler.output_text(unicode=True, color=True))
UnicodeEncodeError: 'ascii' codec can't encode characters in position 345-346: ordinal not in range(128)
```

KSchopmeyer commented:
Fixed all issues except:

1. hide the xml creator.  That is difficult since we are doing multiple tests. we would have to do all building and then all tests and some of this xml is really large.

2. Did not fix the exception issue.  I cannot create it with python 3 and will test with python 2 in morning.

andy-maier commented:
On your comments:
1. We need to hide this. VCreating the XML string likely uses many of the small functions tupleparse is also using (lots of string formatting, isinstance, etc) so not doing this is probably changing the results of the measurement. Plus, it will let improvements appear to be much smaller.
2. I verified that the issue shows up on Python 2, and that that the change fixes it. I also verified that the issue does not show up on Python 3, and that it still works with the change.

KSchopmeyer commented:
Modified the  code to: 

1. Changed name of options for the -p options from stack, table to pyinst, cprofile

2. Changed the long name of the --profiletype to --profiler

3. Changed to build all of the XML responses objects at once.  This may  not have been necessary because of other changes but already done.

4. Added option (-m, --individual) to run a completely separate test for each set of response_count, response_size options.  This outputs any profile information and also the table for each subtest.  NOTE: This has a problem in that right now it outputs the data to the same *.profile name for each individual test.

5. Moved code so that the cprofile runs the profiler ONLY during the execution of the test code.  We are not yet sure that works with the  pyinstrument profiler since, while it has a start and stop, I have not  figured out  if that is really cumulative.

KSchopmeyer commented:
NOTE: I have fixed the print_stats unicode problem yet.

andy-maier commented:
Feedback on the `-i` option:

```
$ tests/manualtest/run_response_performance.py -i
Traceback (most recent call last):
  File "tests/manualtest/run_response_performance.py", line 517, in <module>
    main()
  File "tests/manualtest/run_response_performance.py", line 511, in main
    execute_individual_tests(args)
  File "tests/manualtest/run_response_performance.py", line 320, in execute_individual_tests
    execute_tests(params, dest=dest)
  File "tests/manualtest/run_response_performance.py", line 337, in execute_tests
    table_rows = execute_raw_tests(params, dest)
  File "tests/manualtest/run_response_performance.py", line 233, in execute_raw_tests
    execution_time = execute_with_time(xml, profiler=profiler)
  File "tests/manualtest/run_response_performance.py", line 215, in execute_with_time
    execute_test_code(xml_string, profiler)
  File "tests/manualtest/run_response_performance.py", line 196, in execute_test_code
    profiler.enable()
AttributeError: 'str' object has no attribute 'enable'
```

```
$ tests/manualtest/run_response_performance.py -i -p cprofile
         218302 function calls (210161 primitive calls) in 0.148 seconds

   Ordered by: internal time, call count
   List reduced from 268 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.008    0.008    0.012    0.012 {__import__}
     8000    0.008    0.000    0.014    0.000 abc.py:128(__instancecheck__)
31949/31640    0.008    0.000    0.022    0.000 {isinstance}
     1405    0.006    0.000    0.035    0.000 tupleparse.py:166(check_node)
 2100/900    0.005    0.000    0.028    0.000 _utils.py:132(_ascii2)
      900    0.005    0.000    0.008    0.000 collections.py:50(__init__)
 2700/200    0.005    0.000    0.017    0.000 copy.py:145(deepcopy)
    14957    0.004    0.000    0.004    0.000 _weakrefset.py:70(__contains__)
     3001    0.004    0.000    0.010    0.000 tupletree.py:115(characters)
 1800/900    0.004    0.000    0.037    0.000 string.py:567(_vformat)
        2    0.004    0.002    0.023    0.011 {built-in method Parse}
        1    0.003    0.003    0.003    0.003 expatreader.py:4(<module>)
     1405    0.003    0.000    0.004    0.000 tupletree.py:96(startElement)
      400    0.003    0.000    0.031    0.000 cim_obj.py:4302(__init__)
   1204/1    0.002    0.000    0.113    0.113 tupleparse.py:2044(parse_any)
      400    0.002    0.000    0.003    0.000 {method 'sub' of '_sre.SRE_Pattern' objects}
      700    0.002    0.000    0.010    0.000 _nocasedict.py:90(__init__)
     2605    0.002    0.000    0.003    0.000 tupleparse.py:103(filter_tuples)
     1700    0.002    0.000    0.002    0.000 {method 'format' of 'str' objects}
 1000/200    0.002    0.000    0.069    0.000 tupleparse.py:322(list_of_matching)


Traceback (most recent call last):
  File "tests/manualtest/run_response_performance.py", line 517, in <module>
    main()
  File "tests/manualtest/run_response_performance.py", line 511, in main
    execute_individual_tests(args)
  File "tests/manualtest/run_response_performance.py", line 320, in execute_individual_tests
    execute_tests(params, dest=dest)
  File "tests/manualtest/run_response_performance.py", line 341, in execute_tests
    table_rows = execute_cprofile_tests(params, dest)
  File "tests/manualtest/run_response_performance.py", line 272, in execute_cprofile_tests
    stats = pstats.Stats(dest, stream=stream)
  File "/usr/local/Cellar/python@2/2.7.15_3/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pstats.py", line 81, in __init__
    self.init(arg)
  File "/usr/local/Cellar/python@2/2.7.15_3/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pstats.py", line 95, in init
    self.load_stats(arg)
  File "/usr/local/Cellar/python@2/2.7.15_3/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pstats.py", line 110, in load_stats
    self.stats = marshal.load(f)
EOFError: EOF read where object expected
```

andy-maier commented:
One more: The description of the `-p` option in the help still mentions the old values:
```
-p {none,pyinst,cprofile}, --profiler {none,pyinst,cprofile}
                        Defines the profile package used for the test.
                           * `stack` uses pyinstrument uses a statistical 
                             capture, and displays a tree of the python stack
                             execution times.
                           * `table` uses cProfile and generates a table of
                             counts.
                           * `none` runs without profiler.
                         Default: none
```

andy-maier commented:
Feedback on the `-l` option: It seems a part of the output goes to stdout, and another part gos to the log file:

```
$ tests/manualtest/run_response_performance.py -p cprofile -l
         4973607 function calls (4799151 primitive calls) in 3.023 seconds

   Ordered by: internal time, call count
   List reduced from 268 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   176000    0.172    0.000    0.281    0.000 abc.py:128(__instancecheck__)
735185/734876    0.169    0.000    0.450    0.000 {isinstance}
    35220    0.149    0.000    0.781    0.000 tupleparse.py:166(check_node)
46200/19800    0.119    0.000    0.580    0.000 _utils.py:132(_ascii2)
    19800    0.109    0.000    0.178    0.000 collections.py:50(__init__)
59400/4400    0.100    0.000    0.400    0.000 copy.py:145(deepcopy)
    74812    0.098    0.000    0.208    0.000 tupletree.py:115(characters)
   325757    0.095    0.000    0.095    0.000 _weakrefset.py:70(__contains__)
39600/19800    0.089    0.000    0.778    0.000 string.py:567(_vformat)
       32    0.082    0.003    0.493    0.015 {built-in method Parse}
    35220    0.061    0.000    0.084    0.000 tupletree.py:96(startElement)
8800/6600    0.057    0.000    0.119    0.000 copy.py:234(_deepcopy_tuple)
     8800    0.057    0.000    0.643    0.000 cim_obj.py:4302(__init__)
  30816/4    0.056    0.000    2.525    0.631 tupleparse.py:2044(parse_any)
    61620    0.053    0.000    0.077    0.000 tupleparse.py:103(filter_tuples)
    15400    0.048    0.000    0.207    0.000 _nocasedict.py:90(__init__)
     8800    0.046    0.000    0.060    0.000 {method 'sub' of '_sre.SRE_Pattern' objects}
22000/4400    0.044    0.000    1.553    0.000 tupleparse.py:322(list_of_matching)
   220099    0.043    0.000    0.043    0.000 {getattr}
     6600    0.043    0.000    1.041    0.000 tupleparse.py:1318(parse_property)
```

```
$ cat perf_20190301_063224_cprofile.log
Results: profile=cprofile, response_counts=[100, 1000],
   response-sizes=[100, 1000], 2019-03-01 06:32:24.110136
+----------------+----------------+------------+--------------+
|   Exp Response |   Act Response |   Response |   Parse time |
|     Size Bytes |   Size (Bytes) |      Count |        sec.) |
+================+================+============+==============+
|            100 |            586 |        100 |     0.127838 |
+----------------+----------------+------------+--------------+
|            100 |            586 |       1000 |     1.28374  |
+----------------+----------------+------------+--------------+
|           1000 |           1036 |        100 |     0.132332 |
+----------------+----------------+------------+--------------+
|           1000 |           1036 |       1000 |     1.41706  |
+----------------+----------------+------------+--------------+
```

andy-maier commented:
Once the issues described above are fixed, I agree that it does what the shell script does, so we can remove the shell script.

andy-maier commented:
One more thing: From the help description, it does not become clear as to whether the list items in the `-c` option and the list items in the `-s` option are multiplied out or must correlate.

andy-maier commented:
Feedback on not using `-l` option: the table at the end does no longer get created when not using the `-l` option:

```
$ tests/manualtest/run_response_performance.py -c 10000 -s 500 -p cprofile >perf2_0_base_table_10000_500.log

$ cat perf2_0_base_table_10000_500.log
         21533320 function calls (20782679 primitive calls) in 13.135 seconds

   Ordered by: internal time, call count
   List reduced from 268 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   800000    0.752    0.000    1.213    0.000 abc.py:128(__instancecheck__)
3160349/3160040    0.701    0.000    1.914    0.000 {isinstance}
40000/30000    0.700    0.000    0.961    0.000 copy.py:234(_deepcopy_tuple)
   140005    0.587    0.000    3.299    0.000 tupleparse.py:166(check_node)
210000/90000    0.560    0.000    2.543    0.000 _utils.py:132(_ascii2)
    90000    0.447    0.000    0.737    0.000 collections.py:50(__init__)
270000/20000    0.434    0.000    2.160    0.000 copy.py:145(deepcopy)
   140005    0.407    0.000    0.499    0.000 tupletree.py:96(startElement)
  1480157    0.402    0.000    0.402    0.000 _weakrefset.py:70(__contains__)
180000/90000    0.378    0.000    3.387    0.000 string.py:567(_vformat)
       92    0.332    0.004    2.059    0.022 {built-in method Parse}
   300007    0.304    0.000    0.755    0.000 tupletree.py:115(characters)
    40000    0.249    0.000    2.741    0.000 cim_obj.py:4302(__init__)
 120004/1    0.216    0.000   11.066   11.066 tupleparse.py:2044(parse_any)
    70000    0.207    0.000    0.861    0.000 _nocasedict.py:90(__init__)
   260005    0.206    0.000    0.301    0.000 tupleparse.py:103(filter_tuples)
    40000    0.195    0.000    0.260    0.000 {method 'sub' of '_sre.SRE_Pattern' objects}
100000/20000    0.194    0.000    6.462    0.000 tupleparse.py:322(list_of_matching)
    30000    0.184    0.000    4.523    0.000 tupleparse.py:1318(parse_property)
   980087    0.179    0.000    0.179    0.000 {getattr}
```

andy-maier commented:
Feedback on the measurement results: I understand that the cprofile profiler now measures only the tupleparse code. Here is the output before that change and after that change. Most of the values are identical. That is a surprise to me. Can you please double check that it really only measures the test code? Can you reduce the test code to something simple that just uses a builtin function, to verify that?

```
$ cat perf_0_base_table_10000_500.log
         21593333 function calls (20842692 primitive calls) in 43.691 seconds

   Ordered by: internal time, call count
   List reduced from 277 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1   29.751   29.751   29.900   29.900 run_response_performance.py:90(create_xml_objs)
   800000    0.779    0.000    1.268    0.000 abc.py:128(__instancecheck__)
3160350/3160041    0.717    0.000    1.984    0.000 {isinstance}
40000/30000    0.707    0.000    0.978    0.000 copy.py:234(_deepcopy_tuple)
   140005    0.596    0.000    3.421    0.000 tupleparse.py:166(check_node)
210000/90000    0.586    0.000    2.658    0.000 _utils.py:132(_ascii2)
    90000    0.468    0.000    0.770    0.000 collections.py:50(__init__)
270000/20000    0.451    0.000    2.229    0.000 copy.py:145(deepcopy)
   140005    0.450    0.000    0.550    0.000 tupletree.py:96(startElement)
  1480157    0.425    0.000    0.425    0.000 _weakrefset.py:70(__contains__)
180000/90000    0.395    0.000    3.537    0.000 string.py:567(_vformat)
       92    0.344    0.004    2.181    0.024 {built-in method Parse}
   300007    0.317    0.000    0.795    0.000 tupletree.py:115(characters)
    40000    0.255    0.000    2.854    0.000 cim_obj.py:4302(__init__)
 120004/1    0.221    0.000   11.481   11.481 tupleparse.py:2044(parse_any)
    70000    0.220    0.000    0.902    0.000 _nocasedict.py:90(__init__)
   260005    0.215    0.000    0.310    0.000 tupleparse.py:103(filter_tuples)
    40000    0.209    0.000    0.275    0.000 {method 'sub' of '_sre.SRE_Pattern' objects}
   180006    0.203    0.000    0.203    0.000 {method 'format' of 'str' objects}
100000/20000    0.197    0.000    6.718    0.000 tupleparse.py:322(list_of_matching)


Results: profile=table response_counts=[10000]
   response-sizes=[500] 2019-02-28 05:56:22.647528
+----------------+----------------+------------+--------------+
|   Exp Response |   Act Response |   Response |   Parse time |
|     Size Bytes |   Size (Bytes) |      Count |        sec.) |
+================+================+============+==============+
|            500 |            589 |      10000 |      13.7772 |
+----------------+----------------+------------+--------------+
```

```
$ cat perf2_0_base_table_10000_500.log
         21533320 function calls (20782679 primitive calls) in 13.135 seconds

   Ordered by: internal time, call count
   List reduced from 268 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   800000    0.752    0.000    1.213    0.000 abc.py:128(__instancecheck__)
3160349/3160040    0.701    0.000    1.914    0.000 {isinstance}
40000/30000    0.700    0.000    0.961    0.000 copy.py:234(_deepcopy_tuple)
   140005    0.587    0.000    3.299    0.000 tupleparse.py:166(check_node)
210000/90000    0.560    0.000    2.543    0.000 _utils.py:132(_ascii2)
    90000    0.447    0.000    0.737    0.000 collections.py:50(__init__)
270000/20000    0.434    0.000    2.160    0.000 copy.py:145(deepcopy)
   140005    0.407    0.000    0.499    0.000 tupletree.py:96(startElement)
  1480157    0.402    0.000    0.402    0.000 _weakrefset.py:70(__contains__)
180000/90000    0.378    0.000    3.387    0.000 string.py:567(_vformat)
       92    0.332    0.004    2.059    0.022 {built-in method Parse}
   300007    0.304    0.000    0.755    0.000 tupletree.py:115(characters)
    40000    0.249    0.000    2.741    0.000 cim_obj.py:4302(__init__)
 120004/1    0.216    0.000   11.066   11.066 tupleparse.py:2044(parse_any)
    70000    0.207    0.000    0.861    0.000 _nocasedict.py:90(__init__)
   260005    0.206    0.000    0.301    0.000 tupleparse.py:103(filter_tuples)
    40000    0.195    0.000    0.260    0.000 {method 'sub' of '_sre.SRE_Pattern' objects}
100000/20000    0.194    0.000    6.462    0.000 tupleparse.py:322(list_of_matching)
    30000    0.184    0.000    4.523    0.000 tupleparse.py:1318(parse_property)
   980087    0.179    0.000    0.179    0.000 {getattr}
```

andy-maier commented:
One more feature request for cprofile:
- The number of rows shown should be configurable

andy-maier commented:
On re.compile():

```
   200006    0.119    0.000    0.333    0.000 re.py:138(match)
   200006    0.117    0.000    0.117    0.000 re.py:230(_compile)
```

andy-maier commented:
There is one issue with the new update:
```
(pywbem27) andreass-mbp-2:pywbem maiera$ tests/manualtest/run_response_performance.py -c 1000 -s 500 -p cprofile
  File "tests/manualtest/run_response_performance.py", line 252
SyntaxError: Non-ASCII character '\xe2' in file tests/manualtest/run_response_performance.py on line 252, but no encoding declared; see http://python.org/dev/peps/pep-0263/ for details
```

It is the dash on line 252, which is some other variant of dash that your editor chose to use:
```
        # ncalls – how many times the function/method has been called (in case
                 ^
```

andy-maier commented:
I have one more feature request:
* Name the cprofilerstats files and the log files consistently.
* Support an option to specify a run ID string that is made part of the cprofilerstats files and the log files, so that users can have some meaningful string in their names. If not specified, it would default to say "default".

Proposal for the file names that implements both of that:
* cprofile stats: `perf_{runid}_{timestamp}.cprofile`
* log file: `perf_{runid}_{timestamp}_{profile}.log`
